### ISSUE TITLE
Netflix doesn't work

### ISSUE THREAD

**Issue by maxwowpow** [Alex313031/Thorium-Win-AVX2#84]
**System Details**
 - OS: Win 11 (MacOS M2 too)
 - Thorium Version 112.0.5615.166 (Official Build) (64-bit) as well 111.0.5563.111 (Official Build) (64-bit)
 - Normal and Incognito mode

**Problem**
Netflix Browser Error Code M7121-1331

**Screenshots**
![2023-05-18 19_19_52-Clipboard](https://github.com/Alex313031/Thorium-AVX2/assets/169885/b1146e60-ac71-47fd-beda-ffe3d84cdc07)

**Additional Notes**
Widevine Content Decryption Module - Version: 4.10.2557.0
Status - Component already up to date

https://bitmovin.com/demos/drm

![2023-05-18 19_24_02-DRM Secure Stream Test using HTML5 Video Player _ Bitmovin - Thorium](https://github.com/Alex313031/Thorium-AVX2/assets/169885/07781d65-5761-4a9c-9c77-770588254180)

**Comment by gz83** [comment]
@maxwowpow 

Does version 112 have this problem?

**Comment by gz83** [comment]
@maxwowpow 

Does it work in non-incognito mode?

Also, please open the chrome://components/ page and check the status of the Widevine Content Decryption Module component to see if it is up to date and if you can get the version number.

**Comment by nicklvh** [comment]
i have this issue too

**Comment by gz83** [comment]
Is there still a problem that Netflix cannot play? If it still exists, please open the following two pages when the problem occurs, and upload all the content under these two pages to here, so as to locate the problem

chrome://media-internals/#cdms

chrome://histograms/#Media.EME

@maxwowpow @nicklvh

**Comment by nicklvh** [comment]
![image](https://github.com/Alex313031/Thorium-AVX2/assets/62725534/723ea7c7-059d-4af9-b68b-1b12442bb795)
![image](https://github.com/Alex313031/Thorium-AVX2/assets/62725534/8da9439e-83ad-4f82-b38b-457fd780deee)
![image](https://github.com/Alex313031/Thorium-AVX2/assets/62725534/a18b0573-94d6-4dff-a7f4-cb7489d353c4)
![image](https://github.com/Alex313031/Thorium-AVX2/assets/62725534/49d2c593-0c94-45e8-835f-bcfa696142c3)

**Comment by gz83** [comment]
@nicklvh 

Is it convenient for you to check the widevine components according to the guidelines in this reply?

https://github.com/Alex313031/Thorium-AVX2/issues/84#issuecomment-1557250742

**Comment by nicklvh** [comment]
> @nicklvh
> 
> Is it convenient for you to check the widevine components according to the guidelines in this reply?
> 
> [#84 (comment)](https://github.com/Alex313031/Thorium-AVX2/issues/84#issuecomment-1557250742)

i dont understand?
![image](https://github.com/Alex313031/Thorium-AVX2/assets/62725534/0b55977b-bc51-4c13-892e-7b549fc0c197)
if this is what ur talking about then here

**Comment by gz83** [comment]
You need to go to the following page to confirm the status of the widevine component
 chrome://components/

@nicklvh

**Comment by gz83** [comment]
@maxwowpow 

Is it convenient for you to follow the instructions in the comments below to troubleshoot?

https://github.com/Alex313031/Thorium-AVX2/issues/84#issuecomment-1574893704

**Comment by gz83** [comment]
There is a suspected exception with the widevine component, I will discuss this with Alex

**Comment by Alex313031** [comment]
@gz83 @maxwowpow @nicklvh We will be releasing M113 for MacOS soon.

You guys can try it and see if it fixes the problem.

@midzer When you build, can you make sure the widevine args are correct.

Also, @maxwowpow why did you file this issue in this repo, if you are on Mac?

**Comment by Alex313031** [comment]
https://github.com/Alex313031/Thorium-Win/issues/48

**Comment by gz83** [comment]
![image](https://github.com/Alex313031/Thorium-AVX2/assets/17492865/15fddc21-e180-4342-88a6-67eb5396466d)

Please open the chrome://flags page and open the three experimental flags shown in the image to see if it solves the problem you are experiencing.

@maxwowpow @nicklvh

**Comment by Alex313031** [comment]
I will be migrating the windows builds to a non-component widevine, and see if it fixes the problem.

**Comment by Alex313031** [comment]
@maxwowpow @nicklvh Try 114 release, it should be fixed now.

**Comment by MulverineX** [comment]
@Alex313031 I tried the latest release on Windows 11, Thorium doesn't start

**Comment by gz83** [comment]
If it is the windows version, only the avx2 version has been released, if your hardware does not support it or there are problems with the system, then the browser will definitely not open properly, the regular avx version for windows system will be released in the near future.

@MulverineX

**Comment by gz83** [comment]
Try uninstalling the existing version and reinstalling it, also there is a problem with widevine in the M114 branch according to the upstream bug report.

By the way, would you mind sending the URL of the netflix video that is not playing properly?

@maxwowpow

**Comment by MulverineX** [comment]
@gz83 it would be any episode/movie on the site. Only the non-drm preview videos load.

**Comment by Alex313031** [comment]
@MulverineX @maxwowpow @gz83 This is pissing me off that it's not working!!

@gz83  Im going to make a debug build and test Netflix in a Windows 11 VM

**Comment by Alex313031** [comment]
@maxwowpow I have watched netflix on linux and windows with both chromium and thorium, as recently as last month.

This is a new issue, and is either the widevine media foundation path is not working, or netflix tightened their restrictions on widevine verified media path. Since thorium is not signed (like google chrome), if this is netflix's fault, then there isn't anything I can do.

**Comment by MulverineX** [comment]
In that event could it be spoofed?

**Comment by etil2jz** [comment]
I'm experiencing the same issue with other DRM websites. Tried to manually install the WidevineCdm plugin, unsuccessful.

**Comment by maxwowpow** [comment]
tested portable version of chrome and it doesn't work either.

anyone know what's the default path for the widevine plugin?
a symlink could be a solution if documented.

`C:\path_to_chrome+cdn <<==>> <your_external_location>`

**Comment by Alex313031** [comment]
@maxwowpow @MulverineX @nicklvh @etil2jz @gz83  Upon further investigation, this is not a bug in Thorium. All Thorium builds will have this issue, and all Chromium builds (like from woolyss) will have this issue.

It is related to Google tightening restrictions with the latest Widevine release, and some streaming sites updating at the same time (as per internal guidance from Google, who makes widevine, no doubt) to now require VMP (Verified Media Path), which requires signing your application with a .sig file from widevine, something very expensive to do.

For example the comment about the default path to widevine or symlink, this is not the issue. Widevine is installing and registering properly.

__Sites that work just fine: Hulu, Apple Music, Paramount Plus.__

__Sites that do not work: Netflix, Spotify.__

If you want to watch Netflix, you will either have to use Chrome or Firefox, or, I made a streaming app based on Castlabs Electron fork, which they offer a VMP signing service. However, this service only works on their fork (which is not open source), so it's not like I could use it to (illegally mind you, even if it did work) sign thorium for free.

Anyway, the app is called Quark Player, and I use it on Windows for Netflix > https://github.com/Alex313031/quark-player

**Comment by MulverineX** [comment]
So someone would have to essentially reverse engineer the compiled Castlabs binaries. Yikes, this is really terrible for the open internet. Pirates will always find a way, all they're doing is making the experience worse for end users. DRM sucks.

**Comment by MulverineX** [comment]
This could be done legally if a patch tool was created that downloaded the Castlabs binary on the fly and patched the target browser, but thats very contrived and would take way more effort than eg. Spigot Buildtools.

**Comment by Alex313031** [comment]
@MulverineX I'm thinking of seeing how their signing python script works. I know it checks for a magic string that is only present in their electron binaries. I tried running the signing script on thorium, and vanilla electron, with it reporting a "binary not valid" error. If I could somehow spoof the string, or even compile it into thorium, it might be possible to do it. But then we are getting into illegal territory. I might see how much it would cost to get a widevine license for thorium, and maybe set up a gofundme or something. I know that brave browser signs their binaries (both signing with widevine, and signing the installers with a microsoft cert). Maybe I could ask them how they went about it. But even though they are open source, they still get quite a bit of money from ad revenue and donations, so they can afford it.

**Comment by vairag22** [comment]
> I might see how much it would cost to get a widevine license for thorium, and maybe set up a gofundme or something. I know that brave browser signs their binaries (both signing with widevine, and signing the installers with a microsoft cert). Maybe I could ask them how they went about it. But even though they are open source, they still get quite a bit of money from ad revenue and donations, so they can afford it.

About 2yrs ago same thing was discussed in the ungoogled-chromium repo. You might want to check it out.
[ungoogled-chromium Issue#1258](https://github.com/ungoogled-software/ungoogled-chromium/issues/1258#issuecomment-826350804)

TLDR; you need to be a **registered company** and pay Google a lot of money to get a VMP cert.

**Comment by Alex313031** [comment]
@maxwowpow @nicklvh @MulverineX I have contacted both Widevine.org directly, and CastLabs (Which makes the Electron fork with Widevine that is signed, that I use for Quark Player and other projects use as well like Apple Music), and the price for a license is above 5,000 $US.

That said, unless Thorium becomes really big (relatively) like Brave or Vivaldi, there's no way I could get enough community funding for this. And it is not something I can pay out of pocket for, considering it's an open source hobby project and I don't even have a car right now.

**Comment by MulverineX** [comment]
> The easiest way to stop piracy is not by putting anti-piracy technology to work. It's by giving those people a service that's better than what they're receiving from the pirates.
Gabe Newell

**Comment by MulverineX** [comment]
> The easiest way to stop piracy is not by putting anti-piracy technology to work. It's by giving those people a service that's better than what they're receiving from the pirates.
Gabe Newell

Gotta love how quickly media company's forgot this lesson.

**Comment by vairag22** [comment]
>the price for a license is above 5,000 $US.

Damn, that's really expensive. Is it a one-time payment thing or yearly fees?

**Comment by maxwowpow** [comment]
irony of irony - I didn't open this to watch netflix myself (I rarely do), but rather make the Thorium better. Anyone has 2-3+ browsers, so the pain is manageable :) Let's close this?

**Comment by kkm** [comment]
![image](https://github.com/Alex313031/Thorium-AVX2/assets/134290/20251614-5267-4813-be3b-fabc4395eb25)
I think there are new chrome with new widevine from google widevine.

![image](https://github.com/Alex313031/Thorium-AVX2/assets/134290/9013fbe2-b39c-4aa1-9fb3-db98df779213)

**Comment by Alex313031** [comment]
@kkm That is the media foundation widevine, which is seperate from the main one. In any case, it is a combination of both Widevines being updated, and some streaming services (like Netflix) requiring VMP (read comments above about that).

**Comment by Lukas-drz** [comment]
> @maxwowpow @MulverineX @nicklvh @etil2jz @gz83 Upon further investigation, this is not a bug in Thorium. All Thorium builds will have this issue, and all Chromium builds (like from woolyss) will have this issue.
> 
> It is related to Google tightening restrictions with the latest Widevine release, and some streaming sites updating at the same time (as per internal guidance from Google, who makes widevine, no doubt) to now require VMP (Verified Media Path), which requires signing your application with a .sig file from widevine, something very expensive to do.
> 
> For example the comment about the default path to widevine or symlink, this is not the issue. Widevine is installing and registering properly.
> 
> **Sites that work just fine: Hulu, Apple Music, Paramount Plus.**
> 
> **Sites that do not work: Netflix, Spotify.**
> 
> If you want to watch Netflix, you will either have to use Chrome or Firefox, or, I made a streaming app based on Castlabs Electron fork, which they offer a VMP signing service. However, this service only works on their fork (which is not open source), so it's not like I could use it to (illegally mind you, even if it did work) sign thorium for free.
> 
> Anyway, the app is called Quark Player, and I use it on Windows for Netflix > https://github.com/Alex313031/quark-player
> 
> **_TLDR_**: Netflix is permanently broken on Chromium and Thorium, due to factors outside of my control. Use Chrome, Firefox, or Quark Player for Netflix.

Is there a version that can be installed and works without VMP ?

**Comment by gz83** [comment]
VMP validation is mandatory and cannot be disguised or bypassed. If Widevine exists, there must be VMP.

@Lukas-drz

**Comment by Lukas-drz** [comment]
Thanks
Then how do VMP is processed ? What do I need to do/wait for it to be on my browser and start using Spotify ?

**Comment by gz83** [comment]
Currently, Spotify can only be used using browsers or software that can pass VMP verification.

@Lukas-drz

**Comment by Lukas-drz** [comment]
Thorium can't pass it ?
I managed to have it working

**Comment by gz83** [comment]
Thorium currently cannot pass VMP verification. If it could pass VMP verification, there would not be so many feedback related to Widevine.

@Lukas-drz

**Comment by Lukas-drz** [comment]
Okay then see that :
![passed](https://github.com/Alex313031/Thorium-Win-AVX2/assets/47811332/c71c0edf-4670-42e6-a6fe-93547cae3175)
![not_passed](https://github.com/Alex313031/Thorium-Win-AVX2/assets/47811332/f2574b30-a73c-482c-a4e6-eb88db6b1563)
I don't know how but sometimes after rebooting the thorium browser, the folder : ~/.config/thorium/WidevineCdm/ is no more empty and contains :

├── 4.10.2710.0
│   ├── LICENSE
│   ├── manifest.fingerprint
│   ├── manifest.json
│   ├── _metadata
│   │   └── verified_contents.json
│   └── _platform_specific
│       └── linux_x64
│           └── libwidevinecdm.so
└── latest-component-updated-widevine-cdm

5 directories, 6 files
And now it works on spotify

**Comment by MulverineX** [comment]
IIRC lots of sites haven't updated to use VMP yet, thats why this issue is specifically titled for Netflix. Widevine support isn't completely broken as you've shown, but all sites are likely to transition to VMP eventually.

**Comment by Lukas-drz** [comment]
I found out that @kkm was right.
You need to go to chrome://components in order to check for updates to install the drm and now after rebooting thorium Spotify just works.

**Comment by Alex313031** [comment]
@MulverineX @Lukas-drz Thats interesting, the new widevine might allow spotify to work.

@gz83 This is not entirely true. Widevine can exist without VMP, that how we can watch stuff on linux, and some stuff like hulu still works on the windows and mac thorium releases.

**Comment by Lukas-drz** [comment]
Would it be possible to enable by default widevine without passing by chrome://components ? Or just by passing an argument on start-up?

**Comment by ghost** [comment]
turns out this is a windows-only issue it works on linux

**Comment by MulverineX** [comment]
Incorrect, I'm assuming you only tested on a site with basic Widevine like Spotify? Sites like Netflix that use VMP will cease to function. If netflix really is working then there must be some A/B testing going on

**Comment by ghost** [comment]
> Incorrect, I'm assuming you only tested on a site with basic Widevine like Spotify? Sites like Netflix that use VMP will cease to function. If netflix really is working then there must be some A/B testing going on

I tested it with both Spotify and Netflix and it works

**Comment by MulverineX** [comment]
@JustMe-png other than using Firefox or Quark Player, no. And if you use Firefox you only get 720p at a shitty bitrate because Netflix made under the table agreements with Microsoft & Google. 

Note: AFAICT the 1080p Netflix browser extension doesn't work anymore, if anyone has a script that works a link would be appreciated.

**Comment by MulverineX** [comment]
@JustMe-png No. And there never will be. DRM is a bitch. Perhaps in the future hackers will make a bypass, but I doubt you want to download your browser from hacking forums. It has to be built into the browser binary, it is impossible to do with a browser extension. Alex has already said no to pursuing that course of action.

**Comment by MulverineX** [comment]
Uses a special closed source version of electron, because of the way that the DRM is signed into the software, it's impossible to extract and add to a different chromium build

**Comment by chaoscreater** [comment]
Netflix works on Brave just fine, which is based on Chromium. They have a Github repo for that. Perhaps there's a way to check how they did it to allow Netflix to work?

**Comment by Alex313031** [comment]
@MulverineX @JustMe-png VMP is not supported on Linux, so usually sites will allow it to play, but reduce the quality. Netflix and Spotify work just fine in Thorium on Linux, but Netflix is limited to 576p, and Spotify is limited to 256Kbps.

**Comment by imide** [comment]
This is fucking bogus. I know it's out of the maintainer's control, but it's still asinine. If I have to go through all of these hurdles to watch things I am paying for, I'm just gonna pirate.

Thank you, Alex and other maintainers, you are doing a fantastic job. (not trying to be sarcastic, I mean this) Thorium is working fantastic! (other than this one hiccup)
